src/Conversion/ONNXToTorch/NN/ConcatOp.cpp

```diff
@@ -0,0 +1,105 @@
+ * All input tensors must have the same shape, except for the dimension 
+ * size of the axis to concatenate on.”
+ *
+ * Attributes:
+ *	axis	::mlir::IntegerAttr	64-bit signed integer attribute
+ *  ONNX axis value is map to dimension in the torch side.
+ *
+ * Operands:
+ *    inputs	tensor of 8-bit/16-bit/32-bit/64-bit unsigned 
+ *    		integer values or tensor of 8-bit/16-bit/32-bit/64-bit 
+ *    		signless integer values or tensor of bfloat16 type values 
+ *    		or tensor of 16-bit/32-bit/64-bit float values or 
+ *    		tensor of string type values or tensor of 1-bit signless 
+ *    		integer values or tensor of complex type with 32-bit/64-bit
+ *    		float elements values or memref of any type values.
+ *    ONNX inputs map to input tensors in torch side.
+ *
+ * Results:
+ * concat_result    tensor of 8-bit/16-bit/32-bit/64-bit unsigned
+ *              integer values or tensor of 8-bit/16-bit/32-bit/64-bit
+ *              signless integer values or tensor of bfloat16 type values
+ *              or tensor of 16-bit/32-bit/64-bit float values or
+ *              tensor of string type values or tensor of 1-bit signless
+ *              integer values or tensor of complex type with 32-bit/64-bit
+ *              float elements values or memref of any type values.
+ */
+using namespace mlir;
+using namespace mlir::torch;
+using namespace mlir::torch::Torch;
+
+
+class ONNXConcatOpToTorchLowering : public ConversionPattern {
+public:
+  ONNXConcatOpToTorchLowering(TypeConverter &typeConverter, 
+	MLIRContext *ctx)
+      : ConversionPattern(
+	typeConverter, ::mlir::ONNXConcatOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+
+    Location loc = op->getLoc();
+    mlir::MLIRContext *context =  op->getContext();
+    ONNXConcatOp op1 = llvm::dyn_cast<ONNXConcatOp>(op);
+    ONNXConcatOpAdaptor adaptor(op1);
+    
+    ValueRange inputs = op1.inputs();
+    auto axisValue = op1.axisAttr();       // ::mlir::IntegerAttr
+    Value  axisVal = rewriter.create<ConstantIntOp>(loc,axisValue);
+    
+    TensorType op_tensor_type = op1.getType().cast<TensorType>();
+    auto resultTy = Torch::ValueTensorType::get(op1.getContext(),
+	op_tensor_type.getShape(), op_tensor_type.getElementType());
+    std::vector<Value> inputArrayValues;
+    for (unsigned int i = 0; i < inputs.size(); i++)
+    {
+      TensorType inputTensorType =
+	inputs[i].getType().cast<TensorType>();
+      auto inputTy = Torch::ValueTensorType::get(context, 
+	inputTensorType.getShape(), inputTensorType.getElementType());
+      auto inputTorchTensor  = 
+	rewriter.create<torch::TorchConversion::FromBuiltinTensorOp>(
+		      loc, inputTy, inputs[i]);
+      inputArrayValues.push_back(inputTorchTensor);
+    }
+    Value inputShapeList = rewriter.create<PrimListConstructOp>(loc, 
+	Torch::ListType::get(inputArrayValues.front().getType()), 
```

Comment:
I think this type is supposed to be the result type. I would have expected it to either be computed from the entire `inputArrayValues`, or to be `resultTy` (also eliminating the cast op below).

Could you provide some tests that show that this works? If it does, please document your reasoning for generating the code like this in a comment

Reply:
Added tests and added comments in #16